### PR TITLE
LPS-177859 Save user selected columns

### PR DESCRIPTION
**Pull request by rafaelfs98**
[LPS-177821 - Setting columns as default](https://issues.liferay.com/browse/LPS-177821)

### CODE REVIEW

**Comment by rafaelfs98** [comment]
ci:test:sf

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import i18n from '../../i18n';
+import Form from '../Form';
+import {Column} from '../Table';
+
+type ManagementToolbarColumnsProps = {
+	columns: Column[];
+	onClose: () => void;
+};
+
+type CollumnsState = {
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:28]:
```suggestion
type ColumnsState = {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+}) => {
+	const [{columns: contextColumns, columnsFixed}, dispatch] = useContext(
+		ListViewContext
+	);
+
+	const columnsNotFixed = columns.filter(
+		({key}) => !columnsFixed.includes(key)
+	);
+
+	const [selectedColumns, setSelectedColumns] = useState<CollumnsState>(
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:44]:
```suggestion
	const [selectedColumns, setSelectedColumns] = useState<ColumnsState>(
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		() => {
+			const newColumns: CollumnsState = {};
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:46]:
```suggestion
			const newColumns: ColumnsState = {};
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	return (
+		<div className="align-content-between d-flex flex-column">
+			<div className="dropdown-header">
+				<p className="font-weight-bold my-2">
+					{i18n.translate('columns')}
+				</p>
+			</div>
+
+			<div className="body-toolbar">
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:68]: I think the name `body-toolbar` is too vague.

Put a classname that emphasizes the behavior

```suggestion
			<div className="body-toolbar">
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+								}
+								value={
+									(selectedColumns[
+										column.key
+									] as unknown) as string
+								}
+							/>
+						);
+					})}
+				</div>
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:90]: Simplify

```suggestion
					{columnsNotFixed.map((column, index) => <
					                 Form.Checkbox
								checked={selectedColumns[column.key]}
								key={index}
								label={column.value}
								onChange={(event) =>
									setSelectedColumns({
										...selectedColumns,
										[column.key]: event.target.checked,
									})
								}
								value={
									(selectedColumns[
										column.key
									] as unknown) as string
								}
							/>
							)}
				</div>
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+				<Form.Divider />
+
+				<ClayButton
+					disabled={disabled}
+					onClick={() => {
+						dispatch({
+							payload: {
+								columns: {
+									...selectedColumns,
+								},
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarColumns.tsx:103]: Simplify

```suggestion
								columns: selectedColumns,
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarFilter.tsx [hunk 7]
```diff
@@ -135,7 +135,7 @@ const FilterBody = ({buttonRef, filterSchema, setPosition}: FilterBody) => {
 				<Form.Divider />
 			</div>
 
-			<div className="body-filters">
+			<div className="body-toolbar">
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarFilter.tsx:138]: find a better name

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarRight.tsx [hunk 8]
```diff
@@ -91,32 +93,35 @@ const ManagementToolbarRight: React.FC<ManagementToolbarRightProps> = ({
 			)}
 
 			{display.columns && (
-				<ClayDropDownWithItems
-					items={columns}
+				<ClayPopover
+					alignPosition="bottom-right"
+					className="body-columns popover-management-toolbar"
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarRight.tsx:98]: Those classnames are too vague

```suggestion
					className="body-columns popover-management-toolbar"
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarRight.tsx [hunk 9]
```diff
@@ -64,10 +66,10 @@ const ManagementToolbarRight: React.FC<ManagementToolbarRightProps> = ({
 	buttons,
 	display = {columns: true},
 	columns,
-	disabled,
 	filterSchema,
 }) => {
 	const [{pin}, dispatch] = useContext(ListViewContext);
+	const [visible, setVisible] = useState(false);
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/components/ManagementToolbar/ManagementToolbarRight.tsx:72]: Would be good to add context to this state, I know it's visible, but basically anything could be visible here.

Maybe: `columnsDropdownVisible` clear than this

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/context/ListViewContext.tsx [hunk 10]
```diff
@@ -243,7 +255,12 @@ const ListViewContextProvider: React.FC<
 	ListViewContextProviderProps & {children: ReactNode; id: string}
 > = ({children, id, ...initialStateProps}) => {
 	const [filterPinnedStorage] = useStorage<ListViewFilter>(
-		(STORAGE_KEYS.LIST_VIEW + id) as STORAGE_KEYS,
+		(STORAGE_KEYS.LIST_VIEW_PIN + id) as STORAGE_KEYS,
+		{consentType: CONSENT_TYPE.NECESSARY, storageType: 'persisted'}
+	);
+
+	const [columnsStorage] = useStorage<ListViewColumns>(
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/context/ListViewContext.tsx:262]: Move this variable to before `filterPinnedStorage`

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/core/Storage.ts [hunk 11]
```diff
@@ -19,7 +19,8 @@ import {CONSENT_TYPE} from '~/util/enum';
 export enum STORAGE_KEYS {
 	COMPARE_RUNS = '@testray/compare-runs',
 	EXPORT_CASE_IDS = '@testray/export-case-ids',
-	LIST_VIEW = '@testray/listview-',
+	LIST_VIEW_PIN = '@testray/listview-pin-',
+	LIST_VIEW_COLUMNS = '@testray/listview-columns-',
```
Comment (kevenleone) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/core/Storage.ts:23]: Sort

```suggestion
        LIST_VIEW_COLUMNS = '@testray/listview-columns-',
	LIST_VIEW_PIN = '@testray/listview-pin-',
```

**Review by kevenleone — CHANGES_REQUESTED** [review]
@rafaelfs98 good job on that!!

I left few comments for you with suggestions.

Take a look!

**Pull request closed without merging**